### PR TITLE
Pull upstream

### PR DESCRIPTION
**Pull request by zuzunker**
Pulling latest upstream after https://github.com/input-output-hk/rust-cardano/pull/704

### CODE REVIEW

**Review by rooooooooob — APPROVED** [review]
lgtm, just mockchain changes.

File: chain-impl-mockchain/src/leadership/genesis/mod.rs [hunk 1]
```diff
@@ -136,6 +136,137 @@ impl GenesisLeaderSelection {
...
+
+        let mut pools = HashMap::<StakePoolId, (SecretKey<Curve25519_2HashDH>, u64, Value)>::new();
+
+        let (big_pool_id, big_pool_vrf_private_key) = make_pool(&mut ledger);
+        pools.insert(
+            big_pool_id.clone(),
+            (big_pool_vrf_private_key, 0, Value(1000)),
+        );
+
+        for _i in 0..10 {
```
Comment (nicarq) [inline chain-impl-mockchain/src/leadership/genesis/mod.rs:212]: you can just do `_` in Rust :) 

https://play.rust-lang.org/?code=fn%20main()%0A%7B%0A%20%20%20%20for%20_%20in%200..10%20%7B%0A%20%20%20%20%20%20%20%20println!(%22Hello%20World%22)%3B%0A%20%20%20%20%7D%0A%7D&version=stable

File: chain-impl-mockchain/src/leadership/genesis/mod.rs [hunk 2]
```diff
@@ -136,6 +136,137 @@ impl GenesisLeaderSelection {
...
+
+        let mut date = ledger.date;
+
+        let mut empty_slots = 0;
+
+        let mut times_selected_small = 0;
+
+        let nr_slots = slots_per_epoch;
+
+        for _i in 0..nr_slots {
```
Comment (nicarq) [inline chain-impl-mockchain/src/leadership/genesis/mod.rs:240]: same as above

File: chain-impl-mockchain/src/leadership/genesis/mod.rs [hunk 3]
```diff
@@ -136,6 +136,137 @@ impl GenesisLeaderSelection {
...
+        (pool_id, pool_vrf_private_key)
+    }
+
+    #[test]
+    #[ignore]
+    pub fn test_phi() {
+        let slots_per_epoch = 200000;
+        let active_slots_coeff = 0.1;
+
+        let (_genesis_hash, mut ledger) = create_initial_fake_ledger(
```
Comment (nicarq) [inline chain-impl-mockchain/src/leadership/genesis/mod.rs:197]: if this is not used, I think you can also just do `_`

Comment (rooooooooob) [reply chain-impl-mockchain/src/leadership/genesis/mod.rs:197]: It's not used. marking it as _genesis_hash is done to both signify it is not used (so it doesn't trip the rustc unused variable linter), while also still documenting what it is we're iterating over.

This PR is also just updating our fork with IOHK's repo (thus this is their code) before we send them a PR back, so we don't want to be making needless changes which would just complicate our PRs in their repo.

All changes in this PR are also in the in-development/experimental shelley module which we are not using yet for the bridge or anything else.

File: chain-impl-mockchain/src/leadership/genesis/mod.rs (hunk 1 above)
Comment (zuzunker) [reply chain-impl-mockchain/src/leadership/genesis/mod.rs:212]: That's not our code, tho )

**Pull request merged**